subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/WorkPropertyMetadata.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.properties;
+
+import org.gradle.api.internal.tasks.PropertySpecFactory;
+import org.gradle.api.internal.tasks.properties.bean.PropertyMetadata;
+
+public interface WorkPropertyMetadata {
+    PropertyMetadata getPropertyMetadata();
+
+    void visitPropertyValue(PropertyValue propertyValue, PropertyVisitor visitor, PropertySpecFactory specFactory, BeanPropertyContext context);
```

Comment:
The only thing special to "work" properties is the `PropertySpecFactory` here. However, that value is actually passed to the `visitProperties()` method where we call this method from, so we are basically just passing it through. That's not necessary.

Instead we should move the visitor method to `PropertyMetadata` and ditch this separate `WorkPropertyMetadata` type.